// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: flyteidl2/task/common.proto

package task

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on NamedParameter with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *NamedParameter) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on NamedParameter with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in NamedParameterMultiError,
// or nil if none found.
func (m *NamedParameter) ValidateAll() error {
	return m.validate(true)
}

func (m *NamedParameter) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Name

	if all {
		switch v := interface{}(m.GetParameter()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, NamedParameterValidationError{
					field:  "Parameter",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, NamedParameterValidationError{
					field:  "Parameter",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetParameter()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return NamedParameterValidationError{
				field:  "Parameter",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return NamedParameterMultiError(errors)
	}

	return nil
}

// NamedParameterMultiError is an error wrapping multiple validation errors
// returned by NamedParameter.ValidateAll() if the designated constraints
// aren't met.
type NamedParameterMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m NamedParameterMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m NamedParameterMultiError) AllErrors() []error { return m }

// NamedParameterValidationError is the validation error returned by
// NamedParameter.Validate if the designated constraints aren't met.
type NamedParameterValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NamedParameterValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NamedParameterValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NamedParameterValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NamedParameterValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NamedParameterValidationError) ErrorName() string { return "NamedParameterValidationError" }

// Error satisfies the builtin error interface
func (e NamedParameterValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNamedParameter.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NamedParameterValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NamedParameterValidationError{}

// Validate checks the field values on FixedRate with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *FixedRate) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on FixedRate with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in FixedRateMultiError, or nil
// if none found.
func (m *FixedRate) ValidateAll() error {
	return m.validate(true)
}

func (m *FixedRate) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Value

	// no validation rules for Unit

	if all {
		switch v := interface{}(m.GetStartTime()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, FixedRateValidationError{
					field:  "StartTime",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, FixedRateValidationError{
					field:  "StartTime",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetStartTime()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return FixedRateValidationError{
				field:  "StartTime",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return FixedRateMultiError(errors)
	}

	return nil
}

// FixedRateMultiError is an error wrapping multiple validation errors returned
// by FixedRate.ValidateAll() if the designated constraints aren't met.
type FixedRateMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m FixedRateMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m FixedRateMultiError) AllErrors() []error { return m }

// FixedRateValidationError is the validation error returned by
// FixedRate.Validate if the designated constraints aren't met.
type FixedRateValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e FixedRateValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e FixedRateValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e FixedRateValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e FixedRateValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e FixedRateValidationError) ErrorName() string { return "FixedRateValidationError" }

// Error satisfies the builtin error interface
func (e FixedRateValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sFixedRate.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = FixedRateValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = FixedRateValidationError{}

// Validate checks the field values on Schedule with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *Schedule) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Schedule with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ScheduleMultiError, or nil
// if none found.
func (m *Schedule) ValidateAll() error {
	return m.validate(true)
}

func (m *Schedule) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for KickoffTimeInputArg

	switch v := m.Expression.(type) {
	case *Schedule_Rate:
		if v == nil {
			err := ScheduleValidationError{
				field:  "Expression",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

		if all {
			switch v := interface{}(m.GetRate()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ScheduleValidationError{
						field:  "Rate",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ScheduleValidationError{
						field:  "Rate",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetRate()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ScheduleValidationError{
					field:  "Rate",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	case *Schedule_CronExpression:
		if v == nil {
			err := ScheduleValidationError{
				field:  "Expression",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}
		// no validation rules for CronExpression
	default:
		_ = v // ensures v is used
	}

	if len(errors) > 0 {
		return ScheduleMultiError(errors)
	}

	return nil
}

// ScheduleMultiError is an error wrapping multiple validation errors returned
// by Schedule.ValidateAll() if the designated constraints aren't met.
type ScheduleMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ScheduleMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ScheduleMultiError) AllErrors() []error { return m }

// ScheduleValidationError is the validation error returned by
// Schedule.Validate if the designated constraints aren't met.
type ScheduleValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ScheduleValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ScheduleValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ScheduleValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ScheduleValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ScheduleValidationError) ErrorName() string { return "ScheduleValidationError" }

// Error satisfies the builtin error interface
func (e ScheduleValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSchedule.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ScheduleValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ScheduleValidationError{}

// Validate checks the field values on TriggerAutomationSpec with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *TriggerAutomationSpec) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on TriggerAutomationSpec with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// TriggerAutomationSpecMultiError, or nil if none found.
func (m *TriggerAutomationSpec) ValidateAll() error {
	return m.validate(true)
}

func (m *TriggerAutomationSpec) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Type

	switch v := m.Automation.(type) {
	case *TriggerAutomationSpec_Schedule:
		if v == nil {
			err := TriggerAutomationSpecValidationError{
				field:  "Automation",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

		if all {
			switch v := interface{}(m.GetSchedule()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, TriggerAutomationSpecValidationError{
						field:  "Schedule",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, TriggerAutomationSpecValidationError{
						field:  "Schedule",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetSchedule()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return TriggerAutomationSpecValidationError{
					field:  "Schedule",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	default:
		_ = v // ensures v is used
	}

	if len(errors) > 0 {
		return TriggerAutomationSpecMultiError(errors)
	}

	return nil
}

// TriggerAutomationSpecMultiError is an error wrapping multiple validation
// errors returned by TriggerAutomationSpec.ValidateAll() if the designated
// constraints aren't met.
type TriggerAutomationSpecMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m TriggerAutomationSpecMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m TriggerAutomationSpecMultiError) AllErrors() []error { return m }

// TriggerAutomationSpecValidationError is the validation error returned by
// TriggerAutomationSpec.Validate if the designated constraints aren't met.
type TriggerAutomationSpecValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e TriggerAutomationSpecValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e TriggerAutomationSpecValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e TriggerAutomationSpecValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e TriggerAutomationSpecValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e TriggerAutomationSpecValidationError) ErrorName() string {
	return "TriggerAutomationSpecValidationError"
}

// Error satisfies the builtin error interface
func (e TriggerAutomationSpecValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sTriggerAutomationSpec.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = TriggerAutomationSpecValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = TriggerAutomationSpecValidationError{}

// Validate checks the field values on NamedLiteral with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *NamedLiteral) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on NamedLiteral with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in NamedLiteralMultiError, or
// nil if none found.
func (m *NamedLiteral) ValidateAll() error {
	return m.validate(true)
}

func (m *NamedLiteral) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Name

	if all {
		switch v := interface{}(m.GetValue()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, NamedLiteralValidationError{
					field:  "Value",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, NamedLiteralValidationError{
					field:  "Value",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetValue()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return NamedLiteralValidationError{
				field:  "Value",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return NamedLiteralMultiError(errors)
	}

	return nil
}

// NamedLiteralMultiError is an error wrapping multiple validation errors
// returned by NamedLiteral.ValidateAll() if the designated constraints aren't met.
type NamedLiteralMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m NamedLiteralMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m NamedLiteralMultiError) AllErrors() []error { return m }

// NamedLiteralValidationError is the validation error returned by
// NamedLiteral.Validate if the designated constraints aren't met.
type NamedLiteralValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e NamedLiteralValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e NamedLiteralValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e NamedLiteralValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e NamedLiteralValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e NamedLiteralValidationError) ErrorName() string { return "NamedLiteralValidationError" }

// Error satisfies the builtin error interface
func (e NamedLiteralValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sNamedLiteral.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = NamedLiteralValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = NamedLiteralValidationError{}

// Validate checks the field values on OutputReferences with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *OutputReferences) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on OutputReferences with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// OutputReferencesMultiError, or nil if none found.
func (m *OutputReferences) ValidateAll() error {
	return m.validate(true)
}

func (m *OutputReferences) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for OutputUri

	// no validation rules for ReportUri

	if len(errors) > 0 {
		return OutputReferencesMultiError(errors)
	}

	return nil
}

// OutputReferencesMultiError is an error wrapping multiple validation errors
// returned by OutputReferences.ValidateAll() if the designated constraints
// aren't met.
type OutputReferencesMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m OutputReferencesMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m OutputReferencesMultiError) AllErrors() []error { return m }

// OutputReferencesValidationError is the validation error returned by
// OutputReferences.Validate if the designated constraints aren't met.
type OutputReferencesValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e OutputReferencesValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e OutputReferencesValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e OutputReferencesValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e OutputReferencesValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e OutputReferencesValidationError) ErrorName() string { return "OutputReferencesValidationError" }

// Error satisfies the builtin error interface
func (e OutputReferencesValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sOutputReferences.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = OutputReferencesValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = OutputReferencesValidationError{}

// Validate checks the field values on Inputs with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *Inputs) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Inputs with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in InputsMultiError, or nil if none found.
func (m *Inputs) ValidateAll() error {
	return m.validate(true)
}

func (m *Inputs) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetLiterals() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, InputsValidationError{
						field:  fmt.Sprintf("Literals[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, InputsValidationError{
						field:  fmt.Sprintf("Literals[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return InputsValidationError{
					field:  fmt.Sprintf("Literals[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	for idx, item := range m.GetContext() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, InputsValidationError{
						field:  fmt.Sprintf("Context[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, InputsValidationError{
						field:  fmt.Sprintf("Context[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return InputsValidationError{
					field:  fmt.Sprintf("Context[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return InputsMultiError(errors)
	}

	return nil
}

// InputsMultiError is an error wrapping multiple validation errors returned by
// Inputs.ValidateAll() if the designated constraints aren't met.
type InputsMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m InputsMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m InputsMultiError) AllErrors() []error { return m }

// InputsValidationError is the validation error returned by Inputs.Validate if
// the designated constraints aren't met.
type InputsValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e InputsValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e InputsValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e InputsValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e InputsValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e InputsValidationError) ErrorName() string { return "InputsValidationError" }

// Error satisfies the builtin error interface
func (e InputsValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sInputs.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = InputsValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = InputsValidationError{}

// Validate checks the field values on Outputs with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *Outputs) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Outputs with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in OutputsMultiError, or nil if none found.
func (m *Outputs) ValidateAll() error {
	return m.validate(true)
}

func (m *Outputs) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetLiterals() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, OutputsValidationError{
						field:  fmt.Sprintf("Literals[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, OutputsValidationError{
						field:  fmt.Sprintf("Literals[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return OutputsValidationError{
					field:  fmt.Sprintf("Literals[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return OutputsMultiError(errors)
	}

	return nil
}

// OutputsMultiError is an error wrapping multiple validation errors returned
// by Outputs.ValidateAll() if the designated constraints aren't met.
type OutputsMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m OutputsMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m OutputsMultiError) AllErrors() []error { return m }

// OutputsValidationError is the validation error returned by Outputs.Validate
// if the designated constraints aren't met.
type OutputsValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e OutputsValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e OutputsValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e OutputsValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e OutputsValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e OutputsValidationError) ErrorName() string { return "OutputsValidationError" }

// Error satisfies the builtin error interface
func (e OutputsValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sOutputs.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = OutputsValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = OutputsValidationError{}
